### PR TITLE
Added mean and sum ops to MetricsLambda

### LINKED ISSUE #414: Add helper ops for metrics arithmetics
Following the discussion from [here](https://github.com/pytorch/ignite/pull/410#issuecomment-457105415)

> I was wondering whether it is possible to introduce a metric that performs mean, Mean, and compact the following piece of code:

```python
    precision = Precision(average=False)
    recall = Recall(average=False)
    F1 = precision * recall * 2 / (precision + recall + 1e-20)
    F1 = MetricsLambda(lambda t: torch.mean(t).item(), F1)
```
> to:
```python
    precision = Precision(average=False)
    recall = Recall(average=False)
    metrics = Mean(precision * recall * 2 / (precision + recall + 1e-20))
```

Idea is to provide some helper classes like `Mean`, `Sum` and others.

cc @amitibo

### PR DESCRIPTION
**Pull request by vfdev-5**
Fixes #414  

Description:
Added two basic helper ops `mean` and `sum`. To enable the below code:
```python
precision = Precision(average=False)
recall = Recall(average=False)
f1 = (precision * recall * 2 / (precision + recall + 1e-20)).mean()

precision = Precision(average=False)
recall = Recall(average=False)
custom_metric = (precision + recall).sum()
```
Check list:
* [x] New tests are added (if a new feature is modified)
* [x] New doc strings: text and/or example code are in RST format
* [ ] Documentation is updated (if required)

### CODE REVIEW

**Comment by amitibo** [comment]
@vfdev-5 I think u should add these methods to Metric. MetricsLambda will inherit it. You can also add dim or axis as arguments to the methods

**Comment by vfdev-5** [comment]
> I think u should add these methods to Metric.

@amitibo I'm not sure that this make sense. A single metric if it needs to be averaged or summed, probably, should have such arguments on initialization. 
Composed metrics are `MetricsLambda` and that's why I added these methods to `MetricsLambda`.
Could you please develop your idea why we need to add these methods to `Metric`.

> You can also add dim or axis as arguments to the methods

Yes, this make sense, I can add.

**Comment by vfdev-5** [comment]
@jasonkriss @alykhantejani what do you think ?

**Comment by amitibo** [comment]
IMHO the problem is remembering what 'average' means, e.g. what is averaged, the samples or labels (in the case of multilabels). It might even be more  confusing in the case of precision as average precision (AP) has a different meaning. Why was the 'average' argument introduced? If the reason was to allow creating complicated MetricsLambda expressions then it might be to set automatically on all MetricsLambda objects except the topmost.

**Comment by vfdev-5** [comment]
I see what you mean about remembering what is averaged: samples or labels. That's why we are trying to keep the docstring up to date. 


> Why was the 'average' argument introduced? 

@amitibo I think the primary reason was to be able to compute precision/recall per class. And secondary I happend such that we can create complicated expressions using precision/recall. 

> If the reason was to allow creating complicated MetricsLambda expressions then it might be to set automatically on all MetricsLambda objects except the topmost.

If you take a look at all available metrics, there are only precision and recall that involves multiclass option and allow to compute values per class. Other metrics are for regression tasks. There are also some accuracy classes. Yes, we can ensure that if ingredient metric has attribute `average` to check if it is `False`. But this is make sense only for `F-beta` and similar metrics. However, probably we could have an expression where `average=True` required... I have no any example for the moment.

**Comment by amitibo** [comment]
There is a delicate balance between adding functionality and keeping things simple to use. I think that in some cases `ignite` is missing this balance. For example I have a feeling that having `Accuracy` handle both the binary, multiclass and multilabel cases implicitly might have made it less intuitive. I had to resort to reading the code in order to understand how to use it. I think that the `average` argument makes it  more confusing.

**Comment by vfdev-5** [comment]
@amitibo your feedback is really appreciated, let's discuss how could we improve it.

If you have access to pytorch slack, we have `pytorch-ignite` channel where we could talk in more fluent manner.

Anyway, what do you think about `average` argument in `sklearn.metrics.precision_score`, is it less confusing ? In our case, what could make it more obvious/transparent ?

**Comment by amitibo** [comment]
The `average` parameter confused me because it sounds like `average precision` (AP). Maybe calling the argument `mean` or `classwise` might have made it clearer. For me at least :-).
Though, I agree that following the standards of `sklearn` is a good advice, so maybe it is better to leave it as it is now.

**Comment by vfdev-5** [comment]
@amitibo yes I agree that we can follow example of `sklearn`.

> Doesn't the mean() method make the average argument redundant? i.e. aren't the following two expressions equivalent?
```python
p1 = Precision(average=True)
p2 = Precision(average=False).mean()
```
Yes, they should be similar.

Actually, from the beginning we decided to introduce `.mean()` as synthetic sugar to easily write some composite metrics. That's why I added `.mean()` to `MetricsLambda`. 
On the other hand, if we are thinking of `.mean()` as something helpful as arithmetic ops and adding it to `Metric` we can have such redundent situations as `Precision(average=False).mean()`.

We need to discuss more on it. @alykhantejani @jasonkriss what do you think ?

**Comment by alykhantejani** [comment]
I'm don't have a strong preference on where/if this should go in but if we do add it to `Metric` then we should still maintain the old API  i.e. the `average` param to `Precision` etc.  but maybe add deprecation warnings?

**Comment by vfdev-5** [comment]
@zasdfgbnm as you implemented metrics arithmetics, i'd like to know what is your opinion on where to implement `mean`, `sum` methods ? I'd appreciate your point of view :)

**Comment by zasdfgbnm** [comment]
I have no idea on whether this should be on Metric or MetricLambda. I didn't see strong pros/cons on either decision.

The most natural way for me is:
```python
import ignite.metrics.functional as F
# F contains all torch operators, not just mean and sum
# these operators works on all metrics, not just metric lambdas
f1 = F.mean(precision * recall * 2 / (precision + recall + 1e-20))
```

possible implementations are:
```python
# ignite/metrics/functional.py
import torch
import torch.nn.functional as F

def __getattr__(self, attr):
    if hasattr(torch, attr):
        fn = getattr(torch, attr)
    elif hasattr(F, attr):
        fn = getattr(F, attr)
    else:
        raise AttributeError('Unknown')
    def wrapper(*args, **kwargs):
        return MetricLambda(fn, *args, **kwargs)
```

**Comment by zasdfgbnm** [comment]
@vfdev-5 My opinion is: if we are planning to support all ops as above, then better on `Metric`, otherwise if we just want to support a few of commonly used operators, then I prefer put them into `MetricLambda`. I don't know why I like this way, but just because this seems "natural" to me. This might be a very personal preference.

**Comment by vfdev-5** [comment]
@zasdfgbnm thank you for your comments ! I agree with your reasoning. For now we can keep these two ops in `MetricsLambda` and move out either in `Metric` or as `functional`. 

To merge this one it remains to add kwargs.

**Comment by vfdev-5** [comment]
Closed in favor of #427
